### DISCUSSION TITLE
Error when using regex with negative lookaheads in RENOVATE_AUTODISCOVER_FILTER

### DISCUSSION THREAD

**Discussion by muellerst-hg** [renovatebot/renovate#32340] [Request Help] [answered]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

gitlab renovate:38

### Please tell us more about your question or problem

Given I run a renovate:38 container
When I execute the following commands:
```
# regex with negative lookahead
export RENOVATE_AUTODISCOVER_FILTER='/^((?!mygroup\/testrepo)mygroup\/)/'
renovate --autodiscover
```
Then I should receive a list of all repos beginning with `mygroup/` except for `mygroup/testrepo`
But instead I end up with the following error:
```
DEBUG: Autodiscovered 1192 repositories
DEBUG: Applying autodiscoverFilter
       "autodiscoverFilter": ["/^((?!mygroup\\/testrepo)mygroup\\/)/"]
FATAL: Fatal error: Failed to parse regex pattern "/^((?!mygroup\/testrepo)mygroup\/)/"
       "err": {
         "message": "Failed to parse regex pattern \"/^((?!mygroup\\/testrepo)mygroup\\/)/\"",
         "stack": "Error: Failed to parse regex pattern \"/^((?!mygroup\\/testrepo)mygroup\\/)/\"\n    at applyFilters (/usr/local/renovate/lib/workers/global/autodiscover.ts:109:15)\n    at autodiscoverRepositories (/usr/local/renovate/lib/workers/global/autodiscover.ts:57:18)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at start (/usr/local/renovate/lib/workers/global/index.ts:169:14)\n    at /usr/local/renovate/lib/renovate.ts:18:22"
       }
DEBUG: Checking file package cache for expired items
DEBUG: Deleted 0 of 0 file cached entries in 1ms
 INFO: Renovate is exiting with a non-zero code due to the following logged errors
       "loggerErrors": [
         {
           "name": "renovate",
           "level": 60,
           "logContext": "41ho8q-rzVNgwzfoUErsr",
           "err": {
             "message": "Failed to parse regex pattern \"/^((?!mygroup\\/testrepo)mygroup\\/)/\"",
             "stack": "Error: Failed to parse regex pattern \"/^((?!mygroup\\/testrepo)mygroup\\/)/\"\n    at applyFilters (/usr/local/renovate/lib/workers/global/autodiscover.ts:109:15)\n    at autodiscoverRepositories (/usr/local/renovate/lib/workers/global/autodiscover.ts:57:18)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at start (/usr/local/renovate/lib/workers/global/index.ts:169:14)\n    at /usr/local/renovate/lib/renovate.ts:18:22"
           },
           "msg": "Fatal error: Failed to parse regex pattern \"/^((?!mygroup\\/testrepo)mygroup\\/)/\""
         }
       ]
```

Find the regex working with  some examples: https://regex101.com/r/s0PHW3/1 

What's wrong with my regex?
What alternatives would you suggest?

**Comment by RahulGautamSingh** [comment]
We use google-re2 internally for regex ... which doesn't support negative lookaheads. You can  do this instead:

``` ['/^mygroup\\/)/', '!/^(mygroup\\/testrepo)/)/'] ``` (could be non-working only using this to showcase the approach)

Filters are applied one by one from start of array so this should work.

**Reply by muellerst-hg** [reply]
documentation says:

> [# DO NOT use commas inside the filter if your are using env or cli variables to configure it.](https://docs.renovatebot.com/self-hosted-configuration/#autodiscoverfilter)


I need to use RENOVATE_AUTODISCOVER_FILTER  as env var

**Reply by RahulGautamSingh** [reply]
It says inside the filter not the array. Try a json string array and it will work

eg. ```RENOVATE_AUTODISCOVER_FILTER=[\"mygroup/*\",\"\!mygroup/testrepo\"]```

**Reply by muellerst-hg** [reply]
Thanks for your suggestions. Unfortunately it does not work as expected, instead, no repos are filtered at all. Docs state: 
> Only use a single negation and don't mix with other filters because all filters are combined with or.

**Reply by RahulGautamSingh** [reply]
Edge of edge case 😄

Only this regex can get the job done I think, but not practical ig:

```
^mygroup\/([^t].*|t[^e].*|te[^s].*|tes[^t].*|test[^r].*|testr[^e].*|testre[^p].*|testrep[^o].*|testrepo.+|t|te|tes|test|testr|testre|testrep)$
```

**Reply by rarkins** [reply]
What about stringifying `RENOVATE_CONFIG`?

**Reply by RahulGautamSingh** [reply]
That part is resolved for now. The issue at hand is the negavtive lookahead and the OR logic used when there are multiple filters, which is preventing the desriable filtering.

**Reply by RahulGautamSingh** [reply, marked as answer]
We need to implement this https://github.com/renovatebot/renovate/issues/21347 @muellerst-hg you got the time?

**Reply by muellerst-hg** [reply]
unfortunately, my JS/TS knowledge is not sufficient

**Reply by RahulGautamSingh** [reply]
I will handle it then, not timeline though 😅
